crud/stats/module.lua

```diff
@@ -0,0 +1,211 @@
+    local finish_time = clock.monotonic()
+    local latency = finish_time - start_time
+
+    local err = nil
+    local status = 'ok'
+    if call_status == false then
+        status = 'error'
+        err = select(1, ...)
+    end
+
+    -- If not `pairs` call, return values `nil, err`
+    -- treated as error case.
+    local second_return_val = select(2, ...)
+    if opts.pairs == false and second_return_val ~= nil then
+        status = 'error'
+        err = second_return_val
+    end
+
+    -- If space not exists, do not build a separate collector for it.
+    -- Call request for non-existing space will always result in error.
+    -- The resulting overhead is insignificant for existing spaces:
+    -- at worst it would be a single excessive check for an instance lifetime.
+    -- If we can't verify space existence because of network errors,
+    -- it is treated as unknown as well.
+    if status == 'error' and registry.is_unknown_space(space_name) then
+        if type(err) == 'table' and type(err.err) == 'string' then
+            local space_not_found_msg = utils.space_doesnt_exist_msg(space_name)
+            if string.find(err.err, space_not_found_msg) ~= nil then
+                registry.observe_space_not_found()
+                goto return_values
+            end
+        end
+
+        -- We can't rely only on parsing error value because space existence
+        -- is not always the first check in request validation.
+        -- Check explicitly if space do not exist.
+        local space = utils.get_space(space_name, vshard.router.routeall())
+        if space == nil then
+            registry.observe_space_not_found()
+            goto return_values
+        end
+    end
+
+    registry.observe(latency, space_name, op, status)
+
+    :: return_values ::
+
+    if call_status == false then
+        error((...), 2)
+    end
+
+    return ...
+end
+
+--- Wrap CRUD operation call to collect statistics
+--
+--  Approach based on `box.atomic()`:
+--  https://github.com/tarantool/tarantool/blob/b9f7204b5e0d10b443c6f198e9f7f04e0d16a867/src/box/lua/schema.lua#L369
+--
+-- @function wrap
+--
+-- @tparam function func
+--  Function to wrap. First argument is expected to
+--  be a space name string. If statistics enabled,
+--  errors are caught and thrown again.
+--
+-- @tparam string op
+--  Label of registry collectors.
+--  Use `require('crud.stats.module').op` to pick one.
+--
+-- @tparam table opts
+--
+--  @tfield boolean pairs
+--   (Optional, default: false) If false, second return value
+--   of wrapped function is treated as error (`nil, err` case).
+--   Since pairs calls return three arguments as generator
+--   and throw errors if needed, use { pairs = true } to
+--   wrap them.
+--
+-- @return First two arguments of wrapped function output.
+--
+function stats.wrap(func, op, opts)
+    dev_checks('function', 'string', { pairs = '?boolean' })
+
+    return function(...)
```

Comment:
Since you anyway consider the first argument as a space_name change function signature to `function(space_name, ...)`

Reply:
Seems fair

Reply:
Changed